stdlib/universe/map.go

```diff
@@ -147,35 +147,52 @@ func (t *mapTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey
 func (t *mapTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
 	// Prepare the functions for the column types.
 	cols := tbl.Cols()
-	err := t.fn.Prepare(cols)
-	if err != nil {
+	if err := t.fn.Prepare(cols); err != nil {
 		// TODO(nathanielc): Should we not fail the query for failed compilation?
 		return err
 	}
-	// Determine keys return from function
-	var properties map[string]semantic.Type
-	var keys []string
-	var on map[string]bool
 
+	// Determine the properties from type inference.
+	// TODO(jsternberg): Type inference doesn't currently get all of the properties
+	// when with is used, so this will likely return an incomplete set that we have
+	// to complete when we read the first column. This pass should get the columns
+	// that are directly referenced which are the ones capable of changing and the
+	// only ones that can be a null value.
+	properties := t.fn.Type().Properties()
+
+	var on map[string]bool
 	return tbl.Do(func(cr flux.ColReader) error {
 		l := cr.Len()
 		for i := 0; i < l; i++ {
 			m, err := t.fn.Eval(i, cr)
 			if err != nil {
-				return errors.Wrap(err, "failed to evaluate map function")
+				return errors.Wrap(err, codes.Inherit, "failed to evaluate map function")
 			}
+
+			// Merge in the types that we may have missed because type inference omitted them.
 			if i == 0 {
-				properties = m.Type().Properties()
-				keys = make([]string, 0, len(properties))
-				for k := range properties {
-					keys = append(keys, k)
+				// Merge in the valid properties from the type.
+				// This will catch all of the non-referenced keys that are
+				// merged in with merge key or with.
+				for k, typ := range m.Type().Properties() {
+					if execute.ConvertFromKind(typ.Nature()) != flux.TInvalid {
```

Comment:
Why isn't this an error?

Reply:
It's because the value could be null and type inference reported a different type. I've modified this so it merges only the keys that weren't found with type inference. I'm not sure if this will catch all instances since type inference can sometimes return a null type when it's not supposed to, but I think it should end up working.

So I also added a todo and told it to merge it regardless if type inference returned null.